File: src/brs/web/api/http/handler/SubmitNonce.java
package brs.web.api.http.handler;

import brs.Account;
import brs.Blockchain;
import brs.Generator;
import brs.crypto.Crypto;
import brs.props.PropertyService;
import brs.props.Props;
import brs.services.AccountService;
import brs.util.Convert;
import brs.web.api.http.ApiServlet;
import brs.web.api.http.common.LegacyDocTag;
import signumj.crypto.SignumCrypto;
import signumj.entity.SignumAddress;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;

import javax.servlet.http.HttpServletRequest;

import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.function.Function;
import java.util.stream.Collectors;

import static brs.web.api.http.common.Parameters.*;


public final class SubmitNonce extends ApiServlet.JsonRequestHandler {

  private final Map<Long, String> passphrases;
  private final boolean allowOtherSoloMiners;
  private final AccountService accountService;
  private final Blockchain blockchain;
  private final Generator generator;
  private final int checkPointHeight;


  public SubmitNonce(PropertyService propertyService, AccountService accountService, Blockchain blockchain, Generator generator) {
    super(new LegacyDocTag[] {LegacyDocTag.MINING}, SECRET_PHRASE_PARAMETER, NONCE_PARAMETER, ACCOUNT_ID_PARAMETER, BLOCK_HEIGHT_PARAMETER, DEADLINE_PARAMETER);
    SignumCrypto signumCrypto = SignumCrypto.getInstance();
    this.passphrases = propertyService.getStringList(Props.SOLO_MINING_PASSPHRASES)
            .stream()
            .collect(Collectors.toMap(passphrase -> signumCrypto.getAddressFromPassphrase(passphrase).getSignedLongId(), Function.identity()));
    this.allowOtherSoloMiners = propertyService.getBoolean(Props.ALLOW_OTHER_SOLO_MINERS);
    this.checkPointHeight = propertyService.getInt(Props.BRS_CHECKPOINT_HEIGHT);

    List<String> soloRecipientPassphrases = propertyService.getStringList(Props.REWARD_RECIPIENT_PASSPHRASES);
    for(String soloRecipient : soloRecipientPassphrases){
      String[] idAndPassphrase = soloRecipient.split(":", 2);
      this.passphrases.put(SignumAddress.fromEither(idAndPassphrase[0]).getSignedLongId(), idAndPassphrase[1]);
    }

    this.accountService = accountService;
    this.blockchain = blockchain;
    this.generator = generator;
  }

  @Override
  protected
  JsonElement processRequest(HttpServletRequest req) {
    String secret = req.getParameter(SECRET_PHRASE_PARAMETER);
    long nonce = Convert.parseUnsignedLong(req.getParameter(NONCE_PARAMETER));

    String accountId = req.getParameter(ACCOUNT_ID_PARAMETER);

    String submissionHeight = Convert.emptyToNull(req.getParameter(BLOCK_HEIGHT_PARAMETER));

    JsonObject response = new JsonObject();

    if (submissionHeight != null) {
      try {
        int height = Integer.parseInt(submissionHeight);
        if (height < checkPointHeight) {
            response.addProperty("result", "Given block height smaller than the check point height");
            return response;
        }
        if (height != blockchain.getHeight() + 1) {
          response.addProperty("result", "Given block height does not match current blockchain height");
          return response;
        }
      } catch (NumberFormatException e) {
        response.addProperty("result", "Given block height is not a number");
        return response;
      }
    }

    if (secret == null || Objects.equals(secret, "")) {
      long accountIdLong;
      try {
        accountIdLong = SignumAddress.fromEither(accountId).getSignedLongId();
      } catch (Exception e) {
        response.addProperty("result", "Missing Passphrase and Account ID is malformed");
        return response;
      }
      if (passphrases.containsKey(accountIdLong)) {
        secret = passphrases.get(accountIdLong);
      } else {
        response.addProperty("result", "Missing Passphrase and account passphrase not in solo mining config");
        return response;
      }
    }

    if (!allowOtherSoloMiners && !passphrases.containsValue(secret)) {
      response.addProperty("result", "This account is not allowed to mine on this node as the whitelist is enabled and it is not whitelisted.");
      return response;
    }

    byte[] secretPublicKey = Crypto.getPublicKey(secret);
    Account secretAccount = accountService.getAccount(secretPublicKey);
    if(secretAccount != null) {
      try {
        verifySecretAccount(accountService, blockchain, secretAccount, Convert.parseUnsignedLong(accountId));
      } catch (Exception e) {
        response.addProperty("result", e.getMessage());
        return response;
      }
    }

    Generator.GeneratorState generatorState = null;
    if(accountId == null || secretAccount == null) {
      generatorState = generator.addNonce(secret, nonce);
    }
    else {
      Account genAccount = accountService.getAccount(Convert.parseUnsignedLong(accountId));
      if(genAccount == null || genAccount.getPublicKey() == null) {
        response.addProperty("result", "Passthrough mining requires public key in blockchain");
      }
      else {
        byte[] publicKey = genAccount.getPublicKey();
        generatorState = generator.addNonce(secret, nonce, publicKey);
      }
    }

    if(generatorState == null) {
      response.addProperty("result", "failed to create generator");
      return response;
    }

    response.addProperty("result", "success");
    response.addProperty("deadline", generatorState.getDeadlineLegacy());

    return response;
  }

  public static void verifySecretAccount(AccountService accountService, Blockchain blockchain, Account secretAccount, long accountId) throws Exception {
    Account genAccount;
    if (accountId != 0) {
      genAccount = accountService.getAccount(accountId);
    }
    else {
      genAccount = secretAccount;
    }

    if (genAccount != null) {
      Account.RewardRecipientAssignment assignment = accountService.getRewardRecipientAssignment(genAccount);
      long rewardId;
      if (assignment == null) {
        rewardId = genAccount.getId();
      } else if (assignment.getFromHeight() > blockchain.getLastBlock().getHeight() + 1) {
        rewardId = assignment.getPrevRecipientId();
      } else {
        rewardId = assignment.getRecipientId();
      }
      if (rewardId != secretAccount.getId()) {
        throw new Exception("Passphrase does not match reward recipient");
      }
    }
    else {
      throw new Exception("Passphrase is for a different account");
    }
  }
  boolean requirePost() {
    return true;
  }
}

File: src/main/java/brs/http/SubmitNonce.kt
package brs.http

import brs.Blockchain
import brs.Generator
import brs.crypto.Crypto
import brs.grpc.handlers.SubmitNonceHandler
import brs.grpc.proto.ApiException
import brs.http.common.Parameters.ACCOUNT_ID_PARAMETER
import brs.http.common.Parameters.BLOCK_HEIGHT_PARAMETER
import brs.http.common.Parameters.NONCE_PARAMETER
import brs.http.common.Parameters.SECRET_PHRASE_PARAMETER
import brs.props.PropertyService
import brs.props.Props
import brs.services.AccountService
import brs.util.Convert
import brs.util.parseUnsignedLong
import burst.kit.crypto.BurstCrypto
import burst.kit.entity.BurstAddress
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import javax.servlet.http.HttpServletRequest


internal class SubmitNonce(propertyService: PropertyService, private val accountService: AccountService, private val blockchain: Blockchain, private val generator: Generator) : APIServlet.JsonRequestHandler(arrayOf(APITag.MINING), SECRET_PHRASE_PARAMETER, NONCE_PARAMETER, ACCOUNT_ID_PARAMETER, BLOCK_HEIGHT_PARAMETER) {

    private val passphrases: Map<Long, String>
    private val allowOtherSoloMiners: Boolean

    init {
        val burstCrypto = BurstCrypto.getInstance()
        this.passphrases = propertyService.get(Props.SOLO_MINING_PASSPHRASES)
                .map { burstCrypto.getBurstAddressFromPassphrase(it).burstID.signedLongId to it}
                .toMap()
        this.allowOtherSoloMiners = propertyService.get(Props.ALLOW_OTHER_SOLO_MINERS)
    }

    internal override fun processRequest(request: HttpServletRequest): JsonElement {
        var secret: String? = request.getParameter(SECRET_PHRASE_PARAMETER)
        val nonce = request.getParameter(NONCE_PARAMETER).parseUnsignedLong()

        val accountId = request.getParameter(ACCOUNT_ID_PARAMETER)

        val submissionHeight = Convert.emptyToNull(request.getParameter(BLOCK_HEIGHT_PARAMETER))

        val response = JsonObject()

        if (submissionHeight != null) {
            try {
                val height = Integer.parseInt(submissionHeight)
                if (height != blockchain.height + 1) {
                    response.addProperty("result", "Given block height does not match current blockchain height")
                    return response
                }
            } catch (e: NumberFormatException) {
                response.addProperty("result", "Given block height is not a number")
                return response
            }

        }

        if (secret == null || secret.isEmpty()) {
            val accountIdLong: Long
            try {
                accountIdLong = BurstAddress.fromEither(accountId).burstID.signedLongId
            } catch (e: Exception) {
                response.addProperty("result", "Missing Passphrase and Account ID is malformed")
                return response
            }

            if (passphrases.containsKey(accountIdLong)) {
                secret = passphrases[accountIdLong]!!
            } else {
                response.addProperty("result", "Missing Passphrase and account passphrase not in solo mining config")
                return response
            }
        }

        if (!allowOtherSoloMiners && !passphrases.containsValue(secret)) {
            response.addProperty("result", "This account is not allowed to mine on this node as the whitelist is enabled and it is not whitelisted.")
            return response
        }

        val secretPublicKey = Crypto.getPublicKey(secret)
        val secretAccount = accountService.getAccount(secretPublicKey)
        if (secretAccount != null) {
            try {
                SubmitNonceHandler.verifySecretAccount(accountService, blockchain, secretAccount, accountId.parseUnsignedLong())
            } catch (e: ApiException) {
                response.addProperty("result", e.message)
                return response
            }

        }

        val generatorState = if (accountId == null || secretAccount == null) {
            generator.addNonce(secret!!, nonce)
        } else {
            val genAccount = accountService.getAccount(accountId.parseUnsignedLong())
            if (genAccount?.publicKey == null) {
                response.addProperty("result", "Passthrough mining requires public key in blockchain")
                return response
            } else {
                val publicKey = genAccount.publicKey!!
                generator.addNonce(secret!!, nonce, publicKey)
            }
        }

        response.addProperty("result", "success")
        response.addProperty("deadline", generatorState.deadline)

        return response
    }

    internal override fun requirePost(): Boolean {
        return true
    }
}
